axum/CHANGELOG.md

```diff
@@ -7,56 +7,308 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   `Router::with_state` and gives compile errors for missing states, whereas
   `Extension` would result in runtime errors ([#1155])
+
+  We recommend migrating from `Extension` to `State` since that is more type
+  safe and faster. That is done by using `Router::with_state` and `State`.
+
+  This setup in 0.5
+
+  ```rust
+  use axum::{routing::get, Extension, Router};
+
+  let app = Router::new()
+      .route("/", get(handler))
+      .layer(Extension(AppState {}));
+  
+  async fn handler(Extension(app_state): Extension<AppState>) {}
+  
+  #[derive(Clone)]
+  struct AppState {}
+  ```
+
+  Becomes this in 0.6 using `State`:
+
+  ```rust
+  use axum::{routing::get, extract::State, Router};
+
+  let app = Router::with_state(AppState {})
+      .route("/", get(handler));
+  
+  async fn handler(State(app_state): State<AppState>) {}
+  
+  #[derive(Clone)]
+  struct AppState {}
+  ```
+
+  If you have multiple extensions you can use fields on `AppState` and implement
+  `FromRef`:
+
+  ```rust
+  use axum::{extract::{State, FromRef}, routing::get, Router};
+
+  let app = Router::with_state(AppState {
+      client: HttpClient {},
+      database: Database {},
+  })
+  .route("/", get(handler));
+  
+  async fn handler(
+      State(client): State<HttpClient>,
+      State(database): State<Database>,
+  ) {}
+  
+  #[derive(Clone)]
+  struct AppState {
+      client: HttpClient,
+      database: Database,
+  }
+  
+  #[derive(Clone)]
+  struct HttpClient {}
+  
+  impl FromRef<AppState> for HttpClient {
+      fn from_ref(state: &AppState) -> Self {
+          state.client.clone()
+      }
+  }
+  
+  #[derive(Clone)]
+  struct Database {}
+  
+  impl FromRef<AppState> for Database {
+      fn from_ref(state: &AppState) -> Self {
+          state.database.clone()
+      }
+  }
+  ```
 - **breaking:** The following types or traits have a new `S` type param
   (`()` by default) which represents the state ([#1155]):
   - `FromRequest`
   - `RequestParts`
   - `Router`
   - `MethodRouter`
   - `Handler`
-- **breaking:** `Router::route` now only accepts `MethodRouter`s created with
-  `get`, `post`, etc ([#1155])
-- **added:** `Router::route_service` for routing to arbitrary `Service`s ([#1155])
+
+  This extractor in 0.5:
+
+  ```rust
+  struct MyExtractor;
+
+  #[async_trait]
+  impl<B> FromRequest<B> for MyExtractor
+  where
+      B: Send,
+  {
+      type Rejection = StatusCode;
+  
+      async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+          // ...
+      }
+  }
+  ```
+
+  Becomes this in 0.6:
+
+  ```rust
+  struct MyExtractor;
+
+  #[async_trait]
+  impl<S, B> FromRequest<S, B> for MyExtractor
+  where
+      S: Send + Sync,
+      B: Send,
+  {
+      type Rejection = StatusCode;
+  
+      async fn from_request(req: &mut RequestParts<S, B>) -> Result<Self, Self::Rejection> {
+          // ...
+      }
+  }
+  ```
+
+- **fixed:** Don't expose internal type names in `QueryRejection` response. ([#1171])
```

Comment:
This was backported and isn't actually a change compared to the latest 0.5.x, right?

Reply:
Yeah I hadn't forwardported (?) the changes from 0.5.x, but I have now!